File: js/states/victoryState.js
"use strict";

var State = require('./state');
var CONFIG = require('../config');
var helpers = require('../utils/helpers');

function VictoryState(game) {
  State.call(this, game);
}

VictoryState.prototype = Object.create(State.prototype);
VictoryState.prototype.constructor = VictoryState;

VictoryState.prototype.enter = function (params) {
  params = params || {};
  this.score = params.score || 0;
  this.levelIndex = params.levelIndex || 1;
  this.hasNext = params.hasNext !== false;
};

VictoryState.prototype.handleInput = function () {
  if (this.game.input.consumeTap()) {
    if (this.hasNext) {
      this.game.changeState('play', {
        levelIndex: this.levelIndex + 1,
        score: this.score
      });
    } else {
      this.game.changeState('menu');
    }
  }
};

VictoryState.prototype.update = function (dt) {};

VictoryState.prototype.draw = function (ctx) {
  ctx.fillStyle = 'rgba(0,0,0,0.7)';
  ctx.fillRect(0, 0, CONFIG.SCREEN_WIDTH, CONFIG.SCREEN_HEIGHT);
  helpers.drawText(ctx, 'VICTORY', 48, CONFIG.COLORS.GREEN,
    CONFIG.SCREEN_WIDTH / 2, CONFIG.SCREEN_HEIGHT / 3);
  helpers.drawText(ctx, '关卡完成', 32, CONFIG.COLORS.WHITE,
    CONFIG.SCREEN_WIDTH / 2, CONFIG.SCREEN_HEIGHT / 2);
  helpers.drawText(ctx, '得分: ' + this.score, 22, CONFIG.COLORS.YELLOW,
    CONFIG.SCREEN_WIDTH / 2, CONFIG.SCREEN_HEIGHT / 2 + 45);
  var hint = this.hasNext ? '点击屏幕进入下一关' : '恭喜通关！点击返回菜单';
  helpers.drawText(ctx, hint, 20, CONFIG.COLORS.LIGHT_GREY,
    CONFIG.SCREEN_WIDTH / 2, CONFIG.SCREEN_HEIGHT * 2 / 3 + 40);
};

module.exports = VictoryState;

File: states/victory_state.py
"""胜利/过关状态。"""
import pygame

import config
from states.state import State
from utils.helpers import draw_text


class VictoryState(State):
    """过关或全通关画面。"""

    def __init__(self, game):
        super().__init__(game)

    def enter(self, **kwargs):
        self.score = kwargs.get("score", 0)
        self.level_index = kwargs.get("level_index", 1)
        self.has_next = kwargs.get("has_next", True)

    def handle_events(self, events):
        for event in events:
            if event.type == pygame.KEYDOWN:
                if event.key in (pygame.K_RETURN, pygame.K_SPACE):
                    if self.has_next:
                        self.game.change_state("play", level_index=self.level_index + 1)
                    else:
                        self.game.change_state("menu")
                elif event.key == pygame.K_ESCAPE:
                    self.game.running = False

    def update(self, dt):
        pass

    def draw(self, surface):
        overlay = pygame.Surface((config.SCREEN_WIDTH, config.SCREEN_HEIGHT))
        overlay.fill(config.COLOR_BLACK)
        overlay.set_alpha(180)
        surface.blit(overlay, (0, 0))
        draw_text(surface, "VICTORY", 48, config.COLOR_GREEN,
                  config.SCREEN_WIDTH // 2, config.SCREEN_HEIGHT // 3)
        draw_text(surface, "关卡完成", 32, config.COLOR_WHITE,
                  config.SCREEN_WIDTH // 2, config.SCREEN_HEIGHT // 2)
        draw_text(surface, "得分: {}".format(self.score), 22, config.COLOR_YELLOW,
                  config.SCREEN_WIDTH // 2, config.SCREEN_HEIGHT // 2 + 45)
        if self.has_next:
            draw_text(surface, "按 Enter 进入下一关", 20, config.COLOR_LIGHT_GREY,
                      config.SCREEN_WIDTH // 2, config.SCREEN_HEIGHT * 2 // 3 + 40)
        else:
            draw_text(surface, "恭喜通关！按 Enter 返回菜单", 20, config.COLOR_LIGHT_GREY,
                      config.SCREEN_WIDTH // 2, config.SCREEN_HEIGHT * 2 // 3 + 40)
